### DISCUSSION TITLE
Direct cast expression for input constructor when use inherit

### DISCUSSION THREAD

**Discussion by rootflood** [dotnet/csharplang#4140] [Language Ideas]
make simpler constructor when using inherit constructors
```c#
using System.IO;

    // parent
    public class DirInfo 
    {
        public DirInfo( string Name, int FilesCount, string FullName)
        {}
    }

    public class DirInfoByAddress : DirInfo // inherit ^^^
    {
        public DirInfoByAddress(string Address)  // get address and make DirectoryInfo class
            :this(new DirectoryInfo(Address)){}

        private DirInfoByAddress(DirectoryInfo Info) : // send DirectoryInfo to parent
            base(Info.Name,
                 Info.GetFiles().Length,
                 Info.FullName){}


        public DirInfoByAddress(string Address)  // use a expression for short
            :base
            {
                Info: new DirectoryInfo(Address)  // <<<
            }
                (Info.Name,
                 Info.GetFiles().Length,
                 Info.FullName){}
    }
```

**Comment by canton7** [comment]
I don't understand what the 

```cs
            {
                Info: new DirectoryInfo(Address)  // <<<
            }
```

syntax does I'm afraid. Could you please add some text which explains your proposal a bit more: what problem does it solve, how does it do this, and what exactly does your new syntax compile to?

**Reply by rootflood** [reply]
i want get directory address and make a DirectoryInfo class and use some of property and pass to parent
so i should have two constructors, one for get address and make DirectoryInfo class, other for use some of property like DirectoryInfo.Length

```c#
    public class DirInfoByAddress : DirInfo
    {
        public DirInfoByAddress(string Address)
            :this(new DirectoryInfo(Address)){}  // make new DirectoryInfo class and pass to other constructor

        private DirInfoByAddress(DirectoryInfo Info)  // using DirectoryInfo property 
            :base(Info.Name,
                 Info.GetFiles().Length,
                 Info.FullName){}
    }
```

i want have just one constructor here
example:

```c#
    public class DirInfoByAddress : DirInfo
    {
        private DirInfoByAddress(string Address)
            :base(new DirectoryInfo(Address).Name,
                 new DirectoryInfo(Address).GetFiles().Length,
                 new DirectoryInfo(Address).FullName){}
    }
```

so proposal is:
should have a block for making things and using that things in base constructor
```c#
    public class DirInfoByAddress : DirInfo // inherit ^^^
    {
        public DirInfoByAddress(string Address)  
            :base
            { Info: new DirectoryInfo(Address) }  // <<<data block for make new inputs
                (Info.Name,
                 Info.GetFiles().Length,
                 Info.FullName){}
    }
```c#

**Reply by HaloFour** [reply]
The two constructors you are currently writing seems much cleaner to me than the syntax that you're suggesting.

There are [other proposals](https://github.com/dotnet/csharplang/discussions/3984) seeking to be able to run code before the call to the base constructor which could also solve your need:

```cs
    public class DirInfoByAddress : DirInfo
    {
        private DirInfoByAddress(string Address) {
            var di = new DirectoryInfo(Address);
            base(di.Name, di.GetFiles().Length, di.FullName);
        }
    }
```

**Reply by rootflood** [reply]
this is almost same but with one problem, this is on code block but if that was in data block can tell you can't write code here, so just data must be here
however if i can write codes before of call base constructor then its better for me
for example:
```c#
        public DirInfoByAddress(string Address)
        {
            if(Directory.Exists(Address) is false)
                Directory.CreateDirectory(Address);
            var di = new DirectoryInfo(Address);
            base(di.Name, di.GetFiles().Length, di.FullName);
        }
```
but i think .net have limitation for write code before of call base constructor, isn't?

**Reply by CyrusNajmabadi** [reply]
> but i think .net have limitation for write code before of call base constructor, isn't?

@halofour is saying that there's an existing proposal which would allow you to do what you want, but in a cleaner fashion. For example, as he linked in his post: https://github.com/dotnet/csharplang/discussions/3984

**Reply by HaloFour** [reply]
@rootflood

> but i think .net have limitation for write code before of call base constructor, isn't?

The limitation .NET enforces is that you can't access `this` or instance members before calling the base constructor, but you can do anything else.  That's how C# supports expressions as arguments to the base constructor.  It's the syntax of the base constructor call that makes it awkward.

**Reply by rootflood** [reply]
i agree ,that syntax is much cleaner than my proposal

**Comment by rootflood** [comment]
https://github.com/dotnet/csharplang/issues/3980
